src/main/java/org/broadinstitute/hellbender/tools/funcotator/FuncotatorDataSourceRetriever.java

```diff
@@ -0,0 +1,221 @@
+            final NioFileCopier xerox = downloadDatasources(GERMLINE_GCLOUD_DATASOURCES_PATH);
+
+            // Validate the data sources if requested:
+            if (validateIntegrity) {
+                logger.info("Integrity validation selected.");
+                validateIntegrity(xerox, GERMLINE_GCLOUD_DATASOURCES_SHA256_PATH);
+            }
+        }
+
+        logger.info("IMPORTANT: You must unzip the downloaded data sources prior to using them with Funcotator.");
+
+        // Token return value:
+        return true;
+    }
+
+    @Override
+    protected void onStartup() {
+        // Make sure the user specified at least one data source to download:
+        if ((!getSomaticDataSources) && (!getGermlineDataSources)) {
+            throw new UserException("Must select either somatic or germline datasources.");
+        }
+
+        if ( overwriteOutputFile ) {
+            logger.info("Overwrite ENABLED.  Will overwrite existing data sources download.");
+        }
+    }
+
+    //==================================================================================================================
+    // Static Methods:
+
+    //==================================================================================================================
+    // Instance Methods:
+
+    private NioFileCopier downloadDatasources(final Path dataSourcesPath) {
+
+        // Get the data sources file:
+        final Path outputDestination = getDataSourceLocalPath(dataSourcesPath);
+
+        // Set up and initiate our download:
+        final NioFileCopier xerox = NioFileCopier.create(dataSourcesPath, outputDestination, overwriteOutputFile);
+        xerox.initiateCopy();
+
+        return xerox;
+    }
+
+    private void validateIntegrity(final NioFileCopier fileCopier, final Path remoteSha256Path) {
+
+        // Get the SHA 256 file:
+        final Path localSha256Path = getSha256File(remoteSha256Path);
+
+        // Read the sha256sum into memory:
+        final String expectedSha256Sum = readSha256SumFromFile(localSha256Path);
+
+        // Calculate the sha256sum of the data sources file:
+        final boolean downloadedDataSourcesAreValid = fileCopier.validateIntegrity(expectedSha256Sum, DigestUtils::sha256Hex);
+
+        // verify the hashes are the same:
+        if ( !downloadedDataSourcesAreValid ) {
+            throw new GATKException("ERROR: downloaded data sources are corrupt!  Unexpected checksum: " + fileCopier.getLatestChecksum() + " != " + expectedSha256Sum);
+        }
+        else {
+            logger.info("Data sources are valid.");
+        }
+    }
+
+    private Path getSha256File(final Path remoteSha256Path) {
+        final Path localSha256Path = getDataSourceLocalPath(remoteSha256Path);
+        logger.info("Retrieving expected checksum file...");
+        NioFileCopier.create(remoteSha256Path, localSha256Path, overwriteOutputFile).setSilentCopy(true).initiateCopy();
```

Comment:
In theory, if the user had an old Sha256 file sitting around but deleted the datasource download it went with, then ran this tool with `overwriteOutputFile == false`, couldn't the integrity check fail even though the download was not actually corrupt? Should you be always overwriting the Sha256 file no matter what (or perhaps download it to a unique temp location to get around this issue)?

Reply:
OBE due to reading the file directly.